File: solutions/csharp/roman-numerals/1/RomanNumerals.cs
using System;
using System.Text;

public static class RomanNumeralExtension
{
    public static string GetRomanThousand(int num) =>
        num switch
        {
            1 => "M",
            2 => "MM",
            3 => "MMM",
            _ => "",
        };
    public static string GetRomanHoundred(int num) =>
        num switch
        {
            1 => "C",
            2 => "CC",
            3 => "CCC",
            4 => "CD",
            5 => "D",
            6 => "DC",
            7 => "DCC",
            8 => "DCCC",
            9 => "CM",
            _ => "",
        };
    public static string GetRomanTenth(int num) =>
        num switch
        {
            1 => "X",
            2 => "XX",
            3 => "XXX",
            4 => "XL",
            5 => "L",
            6 => "LX",
            7 => "LXX",
            8 => "LXXX",
            9 => "XC",
            _ => "",
        };
    public static string GetRomanNumber(int num) =>
        num switch
        {
            1 => "I",
            2 => "II",
            3 => "III",
            4 => "IV",
            5 => "V",
            6 => "VI",
            7 => "VII",
            8 => "VIII",
            9 => "IX",
            _ => "",
        };

    public static string ToRoman(this int value)
    {
        StringBuilder stringBuilder = new();

        int currentValue = value;
        if (currentValue >= 1000)
        {
            stringBuilder.Append(GetRomanThousand(currentValue / 1000));
            currentValue %= 1000;
        }
        if (currentValue >= 100)
        {
            stringBuilder.Append(GetRomanHoundred(currentValue / 100));
            currentValue %= 100;
        }
        if (currentValue >= 10)
        {
            stringBuilder.Append(GetRomanTenth(currentValue / 10));
            currentValue %= 10;
        }
        if (currentValue >= 1)
        {
            stringBuilder.Append(GetRomanNumber(currentValue));
        }

        return stringBuilder.ToString();
    }
}

File: solutions/cpp/roman-numerals/1/roman_numerals.cpp
#include "roman_numerals.h"
#include <array>

namespace roman {
const std::array one{"",  "I",  "II",  "III",  "IV",
                     "V", "VI", "VII", "VIII", "IX"};
const std::array ten = {"",  "X",  "XX",  "XXX",  "XL",
                        "L", "LX", "LXX", "LXXX", "XC"};
const std::array hundred = {"",  "C",  "CC",  "CCC",  "CD",
                            "D", "DC", "DCC", "DCCC", "CM"};
const std::array thousand = {"", "M", "MM", "MMM"};
}; // namespace roman

namespace roman_numerals {
std::string convert(int num) {
  if (num <= 0 || num > 3999) {
    return "";
  }
  std::string result{};
  if (num >= 1000) {
    result += roman::thousand.at(num / 1000);
    num %= 1000;
  }
  if (num >= 100) {
    result += roman::hundred.at(num / 100);
    num %= 100;
  }
  if (num >= 10) {
    result += roman::ten.at(num / 10);
    num %= 10;
  }
  if (num >= 1) {
    result += roman::one.at(num);
  }
  return result;
}
} // namespace roman_numerals
